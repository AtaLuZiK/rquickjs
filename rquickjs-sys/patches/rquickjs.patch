diff --git a/quickjs.c b/quickjs.c
index a39ff8f..477de31 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -2108,6 +2108,14 @@ void JS_FreeRuntime(JSRuntime *rt)
     }
 }
 
+#ifdef CONFIG_PARALLEL
+void JS_ResetCtxStack(JSContext *ctx)
+{
+     ctx->rt->stack_top = js_get_stack_pointer();
+}
+#endif
+
+
 JSContext *JS_NewContextRaw(JSRuntime *rt)
 {
     JSContext *ctx;
@@ -27155,6 +27163,26 @@ void JS_SetModuleLoaderFunc(JSRuntime *rt,
     rt->module_loader_opaque = opaque;
 }
 
+#ifdef CONFIG_MODULE_EXPORTS
+/* Hooks into module loading functions */
+int JS_GetModuleExportEntriesCount(JSModuleDef *m) {
+    return m->export_entries_count;
+}
+
+JSValue JS_GetModuleExportEntry(JSContext *ctx, JSModuleDef *m, int idx) {
+    if (idx >= m->export_entries_count || idx < 0)
+        return JS_UNDEFINED;
+    return JS_DupValue(ctx, m->export_entries[idx].u.local.var_ref->value);
+}
+
+JSAtom JS_GetModuleExportEntryName(JSContext *ctx, JSModuleDef *m, int idx) {
+    if (idx >= m->export_entries_count || idx < 0)
+        return JS_ATOM_NULL;
+    return JS_DupAtom(ctx, m->export_entries[idx].export_name);
+}
+#endif
+
+
 /* default module filename normalizer */
 static char *js_default_module_normalize_name(JSContext *ctx,
                                               const char *base_name,
diff --git a/quickjs.h b/quickjs.h
index 92457a2..b55d4c5 100644
--- a/quickjs.h
+++ b/quickjs.h
@@ -352,6 +352,11 @@ JSRuntime *JS_GetRuntime(JSContext *ctx);
 void JS_SetClassProto(JSContext *ctx, JSClassID class_id, JSValue obj);
 JSValue JS_GetClassProto(JSContext *ctx, JSClassID class_id);
 
+#ifdef CONFIG_PARALLEL
+void JS_ResetCtxStack(JSContext *ctx);
+#endif
+
+
 /* the following functions are used to select the intrinsic object to
    save memory */
 JSContext *JS_NewContextRaw(JSRuntime *rt);
@@ -1034,6 +1039,13 @@ int JS_SetModuleExport(JSContext *ctx, JSModuleDef *m, const char *export_name,
                        JSValue val);
 int JS_SetModuleExportList(JSContext *ctx, JSModuleDef *m,
                            const JSCFunctionListEntry *tab, int len);
+                           
+#ifdef CONFIG_MODULE_EXPORTS
+int JS_GetModuleExportEntriesCount(JSModuleDef *m);
+JSValue JS_GetModuleExportEntry(JSContext *ctx, JSModuleDef *m, int idx);
+JSAtom JS_GetModuleExportEntryName(JSContext *ctx, JSModuleDef *m, int idx);
+#endif
+
 
 #undef js_unlikely
 #undef js_force_inline
