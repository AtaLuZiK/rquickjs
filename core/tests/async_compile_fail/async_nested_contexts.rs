use rquickjs::{async_with, AsyncContext, AsyncRuntime};

pub async fn test() {
    let rt = AsyncRuntime::new().unwrap();
    let ctx_1 = AsyncContext::full(&rt).await.unwrap();
    let ctx_2 = AsyncContext::full(&rt).await.unwrap();
    async_with!(ctx_1 => |ctx_1|{
        async_with!(ctx_2 => |ctx_2|{
            ctx_1.globals().set("t", ctx_2.globals());
        }).await
    })
    .await
}

fn main() {}
