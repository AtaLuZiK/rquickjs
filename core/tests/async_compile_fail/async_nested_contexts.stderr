error[E0521]: borrowed data escapes outside of closure
  --> tests/async_compile_fail/async_nested_contexts.rs:9:38
   |
7  | /      async_with!(ctx_1 => |ctx_1|{
8  | |/         async_with!(ctx_2 => |ctx_2|{
9  | ||             ctx_1.globals().set("t", ctx_2.globals());
   | ||                                      ^^^^^^^^^^^^^^^ `ctx_2` escapes the closure body here
10 | ||         }).await
   | ||__________- `ctx_2` is a reference that is only valid in the closure body
11 | |      })
   | |_______- `ctx_1` declared here, outside of the closure body

error[E0521]: borrowed data escapes outside of closure
  --> tests/async_compile_fail/async_nested_contexts.rs:9:13
   |
7  | /     async_with!(ctx_1 => |ctx_1|{
8  | |         async_with!(ctx_2 => |ctx_2|{
9  | |             ctx_1.globals().set("t", ctx_2.globals());
   | |             ^^^^^^^^^^^^^^^
   | |             |
   | |             `ctx_1` escapes the closure body here
   | |             argument requires that `'1` must outlive `'static`
10 | |         }).await
11 | |     })
   | |      -
   | |      |
   | |______`ctx_1` is a reference that is only valid in the closure body
   |        has type `Ctx<'1>`
